### PR TITLE
Add GetPort to URL

### PR DESCRIPTION
**Pull request by vhvb1989**
fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/896

### CODE REVIEW

File: sdk/core/azure-core/test/ut/url.cpp [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    EXPECT_PRED2(
+        [](int expectedValue, int code) { return expectedValue == code; }, url.GetPort(), expected);
+  }
+
+  TEST(URL, getPortInvalidArg)
+  {
+    EXPECT_THROW(Http::Url url("http://test.com:ThisIsNotAPort"), std::invalid_argument);
+  }
+
+  TEST(URL, getPortOutfRange)
```
Comment (RickWinter) [inline sdk/core/azure-core/test/ut/url.cpp:222]: Valid ports are 0-65535.  Add a test case to check the boundary condition, 65536 is an invalid port

File: sdk/core/azure-core/test/ut/url.cpp [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        [](int expectedValue, int code) { return expectedValue == code; }, url.GetPort(), expected);
+
+    url.SetPort(90);
+    expected = 90;
+
+    EXPECT_PRED2(
+        [](int expectedValue, int code) { return expectedValue == code; }, url.GetPort(), expected);
+  }
+
+  TEST(URL, getPortInvalidArg)
```
Comment (RickWinter) [inline sdk/core/azure-core/test/ut/url.cpp:217]: Add a test case that passes ports as a mix of alpha and numeric.  A good example would be if something was in hex.  We should fail on that,  i.e. 1A11 is an invalid port

**Review by RickWinter — APPROVED** [review]
Add the additional test cases before submitting.

File: sdk/core/azure-core/test/ut/url.cpp (hunk 1 above)
Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/url.cpp:222]: I think we don't have any validation for that or negative numbers. 
Following the SDK rules of "do not validate input data for the request", should we avoid preventing someone to add any number (including negative) for the port?

Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/url.cpp:222]: I think we should try to follow at least some "domain specific" constraints and validation.

File: sdk/core/azure-core/CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,7 @@
 ### New Features
 - Added `strings.hpp` with `Azure::Core::Strings::LocaleInvariantCaseInsensitiveEqual` and `Azure::Core::Strings::ToLower`.
 - Added `OperationCanceledException`.
+- Added `GetPort` to Url.
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CHANGELOG.md:13]:
```suggestion
- Added `GetPort()` to `Url`.
```

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 4]
```diff
@@ -298,6 +298,8 @@ namespace Azure { namespace Core { namespace Http {
     /**
      * @brief Set URL port.
      *
+     * @remark If the port was not set for the url, the returned port is -1.
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:301]: This remark belongs on the `Get` not the `Set`.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 5]
```diff
@@ -384,6 +386,13 @@ namespace Azure { namespace Core { namespace Http {
      */
     const std::string& GetPath() const { return m_encodedPath; }
 
+    /**
+     * @brief Get the port number set for the url.
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:390]: docs consistency nit here/elsewhere - capitalize URL:
```suggestion
     * @brief Get the port number set for the URL.
```

File: sdk/core/azure-core/test/ut/CMakeLists.txt [hunk 6]
```diff
@@ -41,6 +41,7 @@ add_executable (
      telemetry_policy.cpp
      transport_adapter.cpp
      transport_adapter_file_upload.cpp
+     url.cpp
```
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/CMakeLists.txt:44]: Good idea splitting these up.

File: sdk/core/azure-core/test/ut/url.cpp [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        "http://test.com/path/path2?query=value");
+
+    EXPECT_NO_THROW(req.GetUrl().AppendPath("path3"));
+    EXPECT_PRED2(
+        [](std::string a, std::string b) { return a == b; },
+        req.GetUrl().GetAbsoluteUrl(),
+        "http://test.com/path/path2/path3?query=value");
+  }
+
+  TEST(URL, getPort)
```
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/url.cpp:181]: Add a GetPort test without setting the port, to verify we return -1 by default. I wonder if we should throw in such cases instead. We probably don't want invalid ports being returned.

File: sdk/core/azure-core/src/http/url.cpp [hunk 8]
```diff
@@ -36,10 +36,25 @@ Url::Url(const std::string& url)
     auto port_ite = std::find_if_not(
         pos + 1, url.end(), [](char c) { return std::isdigit(static_cast<unsigned char>(c)); });
     m_port = std::stoi(std::string(pos + 1, port_ite));
+
+    // stoi will throw out_of_range when `int` is overflow, but we need to throw if uint16 is
+    // overflow
+    if (m_port > 65535)
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/url.cpp:42]: Good comment.

Use named constant instead of "magic number"

File: sdk/core/azure-core/src/http/url.cpp [hunk 9]
```diff
@@ -36,10 +36,25 @@ Url::Url(const std::string& url)
...
+    }
     pos = port_ite;
   }
 
-  if (pos != url.end() && *pos == '/')
+  if (pos != url.end() && (*pos != '/') && (*pos != '?'))
+  {
+    // only char `\` or `?` is valid after the port (or the end of the url). Any other char is an
+    // invalid input
+    throw std::invalid_argument("The port number contains invalid input.");
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/url.cpp:54]:
```suggestion
    throw std::invalid_argument("The port number contains invalid characters.");
```

File: sdk/core/azure-core/src/http/url.cpp (hunk 8 above)
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/url.cpp:42]: We need such a check on `SetPort` as well.

File: sdk/core/azure-core/test/ut/url.cpp (hunk 7 above)
Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/url.cpp:181]: Think I already added a test for that.
I wouldn't like to throw because is more expensive than checking for -1.
This `GetPort()` api is usually expected for transport adapters only. They need to know if a specific port was set for the url or not.
They will check for != -1 without adding a try/catch for that

Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/url.cpp:181]: Good point.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 10]
```diff
@@ -384,6 +384,15 @@ namespace Azure { namespace Core { namespace Http {
      */
     const std::string& GetPath() const { return m_encodedPath; }
 
+    /**
+     * @brief Get the port number set for the URL.
+     *
+     * @remark If the port was not set for the url, the returned port is -1.
+     *
+     * @return The port number from the url.
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:392]: Please make the `URL` fix everywhere. I didn't comment for all the places.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 11]
```diff
@@ -384,6 +384,15 @@ namespace Azure { namespace Core { namespace Http {
...
+     */
+    uint16_t GetPort() const { return m_port; }
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:394]: This shouldn't be unsigned int, when we are returning -1. Leave it as int.

Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:396]: I have changed this -1 behavior.  We can use 0 instead. Since 0 is an invalid port, we can use it to keep the 'not-defined' port.  A request can't be performed to a port 0.
This help us to document the port max number as well

Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:396]: Do we use `uint16_t` or unsigned ints in general anywhere else in the SDK public surface area?

Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:396]: Are we sure port 0 is valid? It looks like it is reserved.

Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:396]: port 0 is not valid. It can't be used to perform any request.

For us, the value 0 is for using the default standard port for the schema.
The problem with using -1 like before, is that, once anyone called `SetPort(uint16_t port)` they can't go back to say `use the default` for that url.
Now using 0, something like this is valid
```cpp
Url u("http://url.com");
u.SetPort(10);

// Send request or use this URL, then set default port
u.SetPort(0);

// Now the http client should pick the universal default for HTTP
```

Using -1 like before, if someone would set 0 as the port would just cause a throw. And customers can't go back to ask for the default port to be used.

Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:396]: > port 0 is not valid. It can't be used to perform any request.

https://www.grc.com/port_0.htm
> "Port Zero" does not officially exist. It is defined as an invalid port number. But valid Internet packets can be formed and sent "over the wire" to and from "port 0" just as with any other ports.

https://en.wikipedia.org/wiki/Port_(computer_networking)#Port_number
> For TCP, port number 0 is reserved and cannot be used, while for UDP, the source port is optional and a value of zero means no port.

https://www.speedguide.net/port.php?port=0
> Port 0 is reserved by IANA, it is technically invalid to use, but possible. It is sometimes used to fingerprint machines, because different operating systems respond to this port in different ways. Some ISPs may block it because of exploits. Port 0 can be used by applications when calling the bind() command to request the next available dynamically allocated source port number.

https://unix.stackexchange.com/a/180502
> Using Port 0 triggers the operating system to search for and allocate the next available port. This avoids you having to hard code a specific port or having to search for an available port.

Do we want the ability to reset to URL port to "default"? Why?

Similarly, do we want the caller to override the port if the url already contained one?
For example:
```C++
Url u("http://url.com:95");
u.SetPort(10); // Why is this allowed? Why would someone want to do this?
```

File: sdk/core/azure-core/test/ut/url.cpp [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+
+namespace Azure { namespace Core { namespace Test {
+
+  TEST(URL, getters)
+  {
+    Http::HttpMethod httpMethod = Http::HttpMethod::Get;
+    Http::Url url("http://test.url.com");
+    Http::Request req(httpMethod, url);
+
+    // EXPECT_PRED works better than just EQ because it will print values in log
```
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/url.cpp:18]: I don't think we need to keep repeating this comment :p

File: sdk/core/azure-core/test/ut/url.cpp [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+  TEST(URL, getPortConst)
+  {
+    Http::Url const url("https://test.com:500");
+    uint16_t expected = 500;
+
+    EXPECT_PRED2(
+        [](uint16_t expectedValue, uint16_t code) { return expectedValue == code; },
+        url.GetPort(),
+        expected);
+  }
```
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/url.cpp:201]: This test seems redundant (its the same as `getPort`), remove?

Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/url.cpp:201]: noup, this is for a const url.  We are testing that GetPort is available for both const and non const url

Comment (ahsonkhan) [reply sdk/core/azure-core/test/ut/url.cpp:201]: Good point. Missed that.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 11 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:396]: > We can create our own rule for the port 0.

I don't know why we would though. This boxes us if some service needs UDP in the future. I don't see a benefit that's worth it. Why don't we keep using `-1` as the "default" and keep port as an int?

> I see a benefit when mutating the port (since all the URL can be mutated anyway -> path, query). I don't see any reason why to block port overriding and allow path and query mutation.

Makes sense. I agree.

**Comment by ahsonkhan** [comment]
Why was this merged with unresolved feedback?

**Comment by ahsonkhan** [comment]
Discussed offline, we will close on the discussion in a separate issue/PR. @vhvb1989 - please file one if needed.
